adam/visualization/positioning.py

```diff
@@ -0,0 +1,426 @@
+from itertools import combinations
+from typing import Mapping, AbstractSet
+from attr import attrs, attrib
+
+import numpy as np
+from numpy import ndarray
+import torch
+import torch.nn as nn
+from torch.nn import Parameter
+import torch.optim as optim
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+
+from immutablecollections import immutabledict, immutableset, ImmutableDict
+from vistautils.preconditions import check_arg
+
+# see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+ORIGIN = torch.tensor([0.0, 0.0, 0.0], dtype=torch.float)  # pylint: disable=not-callable
+COLLISION_PENALTY = 10
+
+
+@attrs(frozen=True, auto_attribs=True)
+class AdamObject:
+    name: str
+
+
+def main() -> None:
+    ball = AdamObject(name="ball")
+    box = AdamObject(name="box")
+
+    cardboard_box = AdamObject(name="cardboardBox")
+    aardvark = AdamObject(name="aardvark")
+    flamingo = AdamObject(name="flamingo")
+
+    positioning_model = AdamObjectPositioningModel.for_objects(
+        immutableset([ball, box, cardboard_box, aardvark, flamingo])
+    )
+    # we will start with an aggressive learning rate
+    optimizer = optim.SGD(positioning_model.parameters(), lr=1.0)
+    # but will decrease it whenever the loss plateaus
+    learning_rate_schedule = ReduceLROnPlateau(
+        optimizer,
+        "min",
+        # decrease the rate if the loss hasn't improved in
+        # 3 epochs
+        patience=3,
+    )
+
+    iterations = 25
+    for _ in range(iterations):
+        loss = positioning_model()
+        print(f"Loss: {loss.item()}")
+        loss.backward()
+
+        optimizer.step()
+        optimizer.zero_grad()
+
+        learning_rate_schedule.step(loss)
+
+        positioning_model.dump_object_positions()
+
+
+@attrs(frozen=True, slots=True)
+class AxisAlignedBoundingBox:
+    """
+    Defines a 3D Box that is oriented to world axes given a center point (3,)
```

Comment:
It is good to have a short single-line description as the first sentence of a docstring because some tools extract and use it.  For example,

```
Defines a 3D box that is oriented to world axes.

This box is defined by a center point of shape (3,)
and a scale (also of shape (3,)),
defines how a unit cube with the given center 
should be scaled in each dimension
to create this box.

For example, (your stuff after e.g.) 
```